legacy/firmware/crypto.c

```diff
@@ -403,21 +457,22 @@ int cryptoMultisigFingerprint(const MultisigRedeemScriptType *multisig,
     if (pubnodes[i]->public_key.size != 33) return 0;
     if (pubnodes[i]->chain_code.size != 32) return 0;
   }
-  // minsort according to pubkey
-  for (uint32_t i = 0; i < n - 1; i++) {
-    for (uint32_t j = n - 1; j > i; j--) {
-      if (memcmp(pubnodes[i]->public_key.bytes, pubnodes[j]->public_key.bytes,
-                 33) > 0) {
-        swap = pubnodes[i];
-        pubnodes[i] = pubnodes[j];
-        pubnodes[j] = swap;
-      }
-    }
+
+  uint32_t pubkeys_order = multisig->has_pubkeys_order
+                               ? multisig->pubkeys_order
+                               : MultisigPubkeysOrder_PRESERVED;
+
+  if (pubkeys_order == MultisigPubkeysOrder_LEXICOGRAPHIC) {
+    // If the order of pubkeys is lexicographic, we don't want the fingerprint
+    // to depend on the order of the pubnodes, so we sort the pubnodes before
+    // hashing.
+    qsort(pubnodes, n, sizeof(HDNodeType *), comparePubnodesLexicographically);
   }
-  // hash sorted nodes
+
   SHA256_CTX ctx = {0};
   sha256_Init(&ctx);
   sha256_Update(&ctx, (const uint8_t *)&(multisig->m), sizeof(uint32_t));
+  sha256_Update(&ctx, (const uint8_t *)&(pubkeys_order), sizeof(uint32_t));
```

Comment:
...that means that this code is wrong, right?
what we want is, pseudocode:
```
MultisigPubkeysOrder pubkeys_order = ...
sha256_Update(multisig->m, sizeof(multisig->m);
sha256_Update(pubkeys_order, sizeof(pubkeys_order));
...
```
etc for this whole block.

Reply:
I believe the code is correct now. The only difference from your pseudocode is that I cast `MultisigPubkeysOrder` to `uint32_t`.

Reply:
my point is that we're magically relying on all the things being u32-sized, with no actual proof in the code (which was the source of the previous bug)

the code is correct now, but could break again in theory

Reply:
I have no problem changing `uint32_t` to `MultisigPubkeysOrder`. However, I don't still understand what you mean by "but could break again in theory". My code doesn't rely on `MultisigPubkeysOrder` to be `uint32_t`.

Reply:
https://github.com/trezor/trezor-firmware/issues/4547

Reply:
Okay, I understand now. I initially thought you were only referring to `pubkeys_order`, but you are also talking about `m`.